### PR TITLE
[FLINK-16581][table] Minibatch deduplication lack state TTL bug fix

### PR DESCRIPTION
**Pull request by lsyldliu**
## What is the purpose of the change
fix the bug of minibatch deduplication lack state TTL, otherwise, it lead to OOM with long running streaming job
## Brief change log
  - support minibatch deduplication keep first row with state TTL
  - support minibatch deduplication keep last row with state TTL

## Verifying this change

This change added tests and can be verified as follows:
  - MiniBatchDeduplicateKeepLastRowFunctionTest#tesKeepLastRowtProcessingTimeTimerWithStateAndGenerateRetraction
  - MiniBatchDeduplicateKeepFirstRowFunctionTest#tesKeepFirstRowtProcessingTimeTimerWithState
## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by lsyldliu** [comment]
@JingsongLi  can you help me review code?

**Comment by lsyldliu** [comment]
@wuchong can you help me review the code?

**Comment by JingsongLi** [comment]
Hi @lsyldliu , we have some discussion on JIRA.

**Comment by lsyldliu** [comment]
> Hi @lsyldliu , we have some discussion on JIRA.

sorry，I didn't noitce that

**Comment by lsyldliu** [comment]
@wuchong CC

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateTtlConfig.java [hunk 1]
```diff
@@ -144,6 +145,21 @@ public String toString() {
 			'}';
 	}
 
+	@Override
+	public boolean equals(Object o){
```
Comment (libenchao) [inline flink-core/src/main/java/org/apache/flink/api/common/state/StateTtlConfig.java]: We don't need to add `equals` method for `StateTtlConfig`, if you want to compare an instance with `StateTtlConfig.DISABLED`, you can just do it like `stateTtlConfig == StateTtlConfig.DISABLED`

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+	public MapBundleFunctionWithStateRetention(long minRetentionTime){
+		this.minRetentionTime = minRetentionTime;
+		stateCleaningEnabled = minRetentionTime > 1;
+		stateTtlConfig = createTtlConfig(minRetentionTime, stateCleaningEnabled);
+	}
+
+	private StateTtlConfig createTtlConfig(long retentionTime, boolean stateCleaningEnabled) {
+		if (stateCleaningEnabled) {
+			checkArgument(retentionTime > 0);
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java]: Add an error message for this check.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+/**
+ * Basic interface for map bundle processing and could cleanup state.
+ *
+ * @param <K>   The type of the key in the bundle map
+ * @param <V>   The type of the value in the bundle map
+ * @param <IN>  Type of the input elements.
+ * @param <OUT> Type of the returned elements.
+ */
+
+public abstract class MapBundleFunctionWithStateRetention<K, V, IN, OUT> extends MapBundleFunction<K, V, IN, OUT> implements CleanupState {
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java]: why do we need to implements `CleanupState`?

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java [hunk 4]
```diff
@@ -71,4 +71,41 @@ public void testKeepFirstRowWithGenerateRetraction() throws Exception {
 		testHarness.close();
 	}
 
+	@Test
+	public void tesKeepFirstRowWithStateTtl() throws Exception{
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: always add a space before `{`

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java [hunk 5]
```diff
@@ -71,4 +71,41 @@ public void testKeepFirstRowWithGenerateRetraction() throws Exception {
...
+		OneInputStreamOperatorTestHarness<BaseRow, BaseRow> testHarness = createTestHarness(func);
+		testHarness.setup();
+		testHarness.open();
+		testHarness.processElement(record("book", 1L, 12));
+		testHarness.processElement(record("book", 2L, 11));
+		// output is empty because bundle not trigger yet.
+		Assert.assertTrue(testHarness.getOutput().isEmpty());
+		testHarness.processElement(record("book", 1L, 13));
+
+		Thread.sleep(30);
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: Do we have any other way to test this, I don't think `Thread.sleep` is a good idea.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java [hunk 6]
```diff
@@ -46,14 +47,18 @@
...
+	public MiniBatchDeduplicateKeepFirstRowFunction(TypeSerializer<BaseRow> typeSerializer, long minRetentionTime) {
+		super(minRetentionTime);
 		this.typeSerializer = typeSerializer;
 	}
 
 	@Override
 	public void open(ExecutionContext ctx) throws Exception {
 		super.open(ctx);
 		ValueStateDescriptor<Boolean> stateDesc = new ValueStateDescriptor<>("existsState", Types.BOOLEAN);
+		if (!stateTtlConfig.equals(StateTtlConfig.DISABLED)){
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java]: use `stateTtlConfig.isEnabled`. We should also correct the implementation in `JoinRecordStateViews`.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java [hunk 7]
```diff
@@ -41,23 +41,28 @@
...
 	private OneInputStreamOperatorTestHarness<BaseRow, BaseRow> createTestHarness(
-			MiniBatchDeduplicateKeepLastRowFunction func)
-			throws Exception {
+		MiniBatchDeduplicateKeepLastRowFunction func)
+		throws Exception {
 		CountBundleTrigger<Tuple2<String, String>> trigger = new CountBundleTrigger<>(3);
 		KeyedMapBundleOperator op = new KeyedMapBundleOperator(func, trigger);
-		return new KeyedOneInputStreamOperatorTestHarness<>(op, rowKeySelector, rowKeySelector.getProducedType());
+		return new KeyedOneInputStreamOperatorTestHarness<>(op, rowKeySelector,
+			rowKeySelector.getProducedType());
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java]: Is there any changes between these lines? It's not encouraged to refactor code sytle among some other changes.  If it is for code style, the later one is not good: 
1) please add an additional indent for new line parameter.
2) please put each parameter in a speparate line if you think the method call is too long. 

Please refer a more detailed code style guideline: https://flink.apache.org/contributing/code-style-and-quality-preamble.html

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/AbstractTtlDecorator.java [hunk 8]
```diff
@@ -40,7 +40,7 @@
 	/** Whether to renew expiration timestamp on state read access. */
 	final boolean updateTsOnRead;
 
-	/** Whether to renew expiration timestamp on state read access. */
+	/** Whether to return expired value if not cleaned up on state read access. */
```
Comment (wuchong) [inline flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/AbstractTtlDecorator.java]: The original Javadoc is correct.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java (hunk 3 above)
Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java]: I don't think we need an abstraction for such a simple member field. A utility to create StateTtlConfig is enough.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java (hunk 5 above)
Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: Does `testHarness.setProcessingTime()` work in this case? cc @Myasuka , can we use `testHarness.setProcessingTime` to test the expiration of state ttl?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java (hunk 2 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java]: Got it

File: flink-core/src/main/java/org/apache/flink/api/common/state/StateTtlConfig.java (hunk 1 above)
Comment (lsyldliu) [reply flink-core/src/main/java/org/apache/flink/api/common/state/StateTtlConfig.java]: Got it, I will remove these code

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java (hunk 3 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/bundle/MapBundleFunctionWithStateRetention.java]: At first, I implemented it with timer, sorry, I forget to remove it.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java (hunk 6 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java]: I see, tks

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java]: By the way, should I correct the implementation in JoinRecordStateViews?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/AbstractTtlDecorator.java (hunk 8 above)
Comment (lsyldliu) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/ttl/AbstractTtlDecorator.java]: ok

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java (hunk 5 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: state ttl use TtlTimeProvider.DEFAULT to get current process time, testHarness.setProcessingTime doesn't work, whether it is appropriate to use TimeUnit.MILLISECONDS.sleep(30) ? Do you have any better advice?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java (hunk 6 above)
Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java]: Yes. Thanks.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java (hunk 7 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java]: sorry, I didn't notice that

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java (hunk 5 above)
Comment (Myasuka) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: Unfortunately, current we always use `TtlTimeProvider.DEFAULT` to create keyed state backend. One way to walk around this:

1. Define a new `MockTtlTimeProvider` which could set current time outside.
1. Define a `TTLMemoryStateBackend` extended from `MemoryStateBackend` which would ignore the given `TtlTimeProvider.DEFAULT` when creating keyed state backend, but use previous newly defined  `MockTtlTimeProvider`.
1. Set the `TTLMemoryStateBackend` to the newly `KeyedOneInputStreamOperatorTestHarness` to use our customized state backend.

Then you could set a new time to `MockTtlTimeProvider` to ensure time has passed time to live.

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: @wuchong , @Myasuka At the moment, KeyedStateBackend created in `StreamTaskStateInitializerImpl#keyedStateBackend` method using `TtlTimeProvider.DEFAULT` explicitly, didn't support pass TtlTimeProvider to it, even though we implement `TTLMemoryStateBackend `, we also can't  set `MockTtlTimeProvider ` to KeyedStateBackend. Unless, we modify relavant code in `StreamTaskStateInitializerImpl`, but I dare not change the code because I can't evaluate the impact, what do you think?

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: Hi @lsyldliu , I think a better way is to make  `StreamTaskStateInitializerImpl` accept a `TtlTimeProvider` constructor parameter instead of hard code `TtlTimeProvider.DEFAULT` in the implementation. Then, we can pass-in the `MockTtlTimeProvider` in `AbstractStreamOperatorTestHarness#createStreamTaskStateManager`. 

What do you think @Myasuka ? Could you help on this?

Comment (Myasuka) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: @wuchong , sure I could help on this to let `TtlTimeProvider` configurable by opening another issue.

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: Thanks @Myasuka . 
@lsyldliu , we may need to hold this PR until `StreamTaskStateInitializerImpl` supports custom `TtlTimeProvider`.

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: OK

**Comment by wuchong** [comment]
Hi @lsyldliu , thanks for the waiting. @Myasuka helped to support state ttl time provider for AbstractStreamOperatorTestHarness (FLINK-16949). Now, you can use `AbstractStreamOperatorTestHarness#setStateTtlProcessingTime` to manually change the processing time for state ttl. Could you update the code?

**Comment by lsyldliu** [comment]
> Hi @lsyldliu , thanks for the waiting. @Myasuka helped to support state ttl time provider for AbstractStreamOperatorTestHarness ([FLINK-16949](https://issues.apache.org/jira/browse/FLINK-16949)). Now, you can use `AbstractStreamOperatorTestHarness#setStateTtlProcessingTime` to manually change the processing time for state ttl. Could you update the code?

Yeah, I`ve updated the code, you can help review it!
Btw, I also want to work for Minibatch Group Aggregateion make it support state TTL like this, and refactor Group Aggregation, Over Aggregateion etc use StateTtlConfig instead of timer which is current implementation, what do you think about it?

**Comment by wuchong** [comment]
Yes. We should do that. Would you like to create issues for that?

**Comment by lsyldliu** [comment]
> Yes. We should do that. Would you like to create issues for that?

Ok, I will create two issues.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java [hunk 9]
```diff
@@ -42,18 +44,25 @@
...
+		stateCleaningEnabled = minRetentionTime > 1;
 		this.typeSerializer = typeSerializer;
 	}
 
 	@Override
 	public void open(ExecutionContext ctx) throws Exception {
 		super.open(ctx);
 		ValueStateDescriptor<Boolean> stateDesc = new ValueStateDescriptor<>("existsState", Types.BOOLEAN);
+		StateTtlConfig ttlConfig = createTtlConfig(minRetentionTime, stateCleaningEnabled);
+		if (ttlConfig.isEnabled()){
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java:63]: `if (ttlConfig.isEnabled()) {`
we always add a space before `{`

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunction.java:63]: got it

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java [hunk 10]
```diff
@@ -113,4 +113,42 @@ public void testWithGenerateRetraction() throws Exception {
...
+
+		testHarness.processElement(record("book", 1L, 10));
+		testHarness.processElement(record("book", 2L, 11));
+		// output is empty because bundle not trigger yet.
+		Assert.assertTrue(testHarness.getOutput().isEmpty());
+		testHarness.processElement(record("book", 1L, 13));
+
+		testHarness.setStateTtlProcessingTime(30);
+		//Incremental cleanup is an eventual clean up, more state access guarantee more expired state cleaned
+		for (long i = 3; i < 30; i++){
```
Comment (libenchao) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java:132]: nit: `) {`

Comment (libenchao) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java:132]: just curious, if no more other state access, old state will be valid when it is accessed?

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java]: sorry, I have updated it.
Yes, Incremental cleanup is eventual cleanup, needs more state access, it keeps a global iterator and might not have reached and cleaned some expired state.

Comment (libenchao) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java]: have you tested it？
I'm asking because in state backend, we can easily check the ttl when we get the state, and even it's not cleaned physically.

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunctionTest.java]: Yeah, I have tested it. When remove these code, the test failed. You can see [FLINK-10473](https://issues.apache.org/jira/browse/FLINK-10473?focusedCommentId=17068492&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17068492) for more explain.

**Comment by lsyldliu** [comment]
@wuchong can you help review the code? is there anything else I need to do? only after you merge this pr can I start the new work, tks.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java [hunk 11]
```diff
@@ -71,4 +71,41 @@ public void testKeepFirstRowWithGenerateRetraction() throws Exception {
...
+		testHarness.processElement(record("book", 1L, 12));
+		testHarness.processElement(record("book", 2L, 11));
+		// output is empty because bundle not trigger yet.
+		Assert.assertTrue(testHarness.getOutput().isEmpty());
+		testHarness.processElement(record("book", 1L, 13));
+
+		testHarness.setStateTtlProcessingTime(30);
+		//Incremental cleanup is an eventual clean up, more state access guarantee more expired state cleaned
+		for (long i = 3; i < 30; i++) {
+			testHarness.processElement(record("book", i, 20));
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java:89]: cc @Myasuka , is this a correct way to trigger state cleanup? 
Is there a better to make sure the expired state cleaned?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecDeduplicate.scala [hunk 12]
```diff
@@ -117,21 +117,23 @@ class StreamExecDeduplicate(
...
       }
       val trigger = AggregateUtil.createMiniBatchTrigger(tableConfig)
       new KeyedMapBundleOperator(
         processFunction,
         trigger)
     } else {
-      val minRetentionTime = tableConfig.getMinIdleStateRetentionTime
-      val maxRetentionTime = tableConfig.getMaxIdleStateRetentionTime
       val processFunction = if (keepLastRow) {
         new DeduplicateKeepLastRowFunction(minRetentionTime, maxRetentionTime, rowTypeInfo,
```
[the remark this answers is not in the document]
Comment (lsyldliu) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecDeduplicate.scala:138]: no problem

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java (hunk 11 above)
Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java:89]: Similar to`TtlStateTestBase#triggerMoreIncrementalCleanupByOtherOps` which is used to trigger the expired state clean eventually.

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java:89]: My concern is that why the number is 30? Is 30 enough?

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java:89]: In this case, it is enough, I just picked it at random, of course, the bigger the better.

Comment (Myasuka) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java:89]: Since we choose `ReturnExpiredIfNotCleanedUp` for better performance, and verifying the data has been physically deleted is a topic related to the implementation of clean up strategy for different state backends. For RocksDB, we have to reply on the compaction taken for sst files containing those keys. For Heap KeyedStateBackend, the records to fetch for cleaning up is related to the key groups (how many state map would existed) in one state backend. 
I think state backend module could guarantee the correctness, and this should be out of scope for SQL module. Once state backend changes the implementation of TTL cleanup strategy, verifying the data has been physically deleted in SQL module might be unstable.
In a nutshell, we would not need to ensure data has been physically deleted here, verifying the state in `MiniBatchDeduplicateKeepFirstRowFunction` has been configured with TTL is enough. Otherwise, if we use `NeverReturnExpired` strategy, things would be much simpler here.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/StateTtlConfigUtil.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Utility to create a {@link StateTtlConfig} object.
+ * */
+public class StateTtlConfigUtil {
+
+	public static StateTtlConfig createTtlConfig(long retentionTime, boolean stateCleaningEnabled) {
+		if (stateCleaningEnabled) {
+			checkArgument(retentionTime > 0, "ttl time must be positive when enabling state cleanup");
```
Comment (Myasuka) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/util/StateTtlConfigUtil.java:33]: `StateTtlConfig` would already check the TTL time larger than 0, this check might not be so useful.

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunction.java [hunk 14]
```diff
@@ -59,6 +64,10 @@ public MiniBatchDeduplicateKeepLastRowFunction(BaseRowTypeInfo rowTypeInfo, bool
 	public void open(ExecutionContext ctx) throws Exception {
 		super.open(ctx);
 		ValueStateDescriptor<BaseRow> stateDesc = new ValueStateDescriptor<>("preRowState", rowTypeInfo);
+		StateTtlConfig ttlConfig = createTtlConfig(minRetentionTime, stateCleaningEnabled);
+		if (ttlConfig.isEnabled()) {
```
Comment (Myasuka) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepLastRowFunction.java:68]: No need to check here, just enable a `StateTtlConfig.DISABLED` is okay.

File: flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java (hunk 11 above)
Comment (libenchao) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java:89]: @Myasuka Thanks for the input.
+1 to use `NeverReturnExpired` strategy. We have got other questions from user ML about this too. IMHO, `NeverReturnExpired` is more straight forward to understand for users.

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: I also discussed with @lincoln-lil and we think `NeverReturnExpired` is better to give a more deterministic result for operators. +1 to use `NeverReturnExpired`. Could you update the strategy @lsyldliu ? And we don't need this foreach then.

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: Hi @libenchao , what's the question in user mailing list?

Comment (libenchao) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: @wuchong a mail titled "关于状态TTL" in user-zh ML.

Comment (wuchong) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: Thanks @libenchao .

Comment (lsyldliu) [reply flink-table/flink-table-runtime-blink/src/test/java/org/apache/flink/table/runtime/operators/deduplicate/MiniBatchDeduplicateKeepFirstRowFunctionTest.java]: Okay, I also prefer to `NeverReturnExpired` strategy, I choose `ReturnExpiredIfNotCleanedUp ` because stream-stream join use it currently, I think it should be unified.

**Comment by lsyldliu** [comment]
@wuchong I have updated state ttl strategy and rebase master branch, you can help review the code now, tks.

**Comment by lsyldliu** [comment]
@wuchong  Thanks for your review and merge, I`m very excited about making a little contribution to the community.
